src/fileformats/ocd_file_export.cpp

```diff
@@ -151,6 +151,68 @@ void copySymbolHead(const Symbol& source, Symbol& symbol)
 
 
 
+/**
+ * Test if the fill patterns constitute OCD structure "shifted rows".
+ */
+bool isStructureShiftedRows(const AreaSymbol::FillPattern& first, const AreaSymbol::FillPattern& second)
+{
+	auto renamed = [](PointSymbol const& symbol, QString const& name) {
+		auto clone = duplicate(symbol);
+		clone->setName(name);
+		return clone;
+	};
+	return first.type == AreaSymbol::FillPattern::PointPattern
+	       && second.type == AreaSymbol::FillPattern::PointPattern
+	       && first.rotatable() == second.rotatable()
+	       && qFuzzyCompare(first.angle, second.angle)
+	       && first.line_spacing == second.line_spacing
+	       && first.point_distance == second.point_distance
+	       && qAbs((first.line_offset - second.line_offset + first.line_spacing) % first.line_spacing - first.line_spacing / 2) <= 1
+	       && qAbs((first.offset_along_line - second.offset_along_line + first.point_distance) % first.point_distance - first.point_distance / 2) <= 1
+	       && first.point
+	       && second.point
+	       && first.point->equals(renamed(*second.point, first.point->getName()).get());
+}
+
+
+
+/**
+ * Postprocess a fill pattern for optimal OCD structure export.
+ */
+std::unique_ptr<PointSymbol> postProcessFillPattern(const AreaSymbol::FillPattern* first_point_pattern)
+{
+	auto point_symbol = duplicate(*first_point_pattern->point);
+	
+	// Move first pattern as needed
+	if (first_point_pattern->line_offset != 0
+	    || first_point_pattern->offset_along_line != 0)
+	{
+		for (int i = 0, last = point_symbol->getNumElements(); i < last; ++i)
+		{
+			if (auto* element = point_symbol->getElementObject(i))
+				element->move(first_point_pattern->line_offset, first_point_pattern->offset_along_line);
+		}
+		if ((point_symbol->getInnerRadius() > 0 && point_symbol->getInnerColor())
+		    || (point_symbol->getOuterWidth() > 0 && point_symbol->getOuterColor()))
+		{
+			// The central dot/circle must be converted to a proper point symbol to apply the offset.
+			auto* primary_element = new PointSymbol();
+			primary_element->setInnerRadius(point_symbol->getInnerRadius());
+			primary_element->setInnerColor(point_symbol->getInnerColor());
+			point_symbol->setInnerColor(nullptr);
+			primary_element->setOuterWidth(point_symbol->getOuterWidth());
+			primary_element->setOuterColor(point_symbol->getOuterColor());
+			point_symbol->setInnerColor(nullptr);
+			auto* object = new PointObject(primary_element);
+			object->setPosition(first_point_pattern->line_offset, first_point_pattern->offset_along_line);
+			point_symbol->addElement(0, object, primary_element);
+		}
+	}
+	return point_symbol;
```

Comment:
Note that I would not call this an "optimization". It is a transformation. The individual pattern elements are transformed so that pattern offset is 0,0 (suitable for OCD), and multiple Mapper patterns are merged into a single OCD pattern (which is all we get from OCD).

Reply:
So I've spent another hour trying to get through the code. Not a complete success yet, but I can see the structure better. The surrounding comments are a result of the effort.

Reply:
So after another cycle through the method, I guess that it handles the case when the Mapper area fill pattern is shifted against the object origin. It needs to be translated for OCD export. Not sure if I've arrived on the spot but it does not matter as I'd like to stress the need of documenting design ideas either in the code comments or commit messages. I prefer the commit message because code comments tend to become outdated while the design idea is forever bound with the actual code change implementing it.